test/io-tests/test_io.py

```diff
@@ -0,0 +1,242 @@
+        try:
+            response = requests.get(url, timeout=0.1)
+
+            if response.status_code == 200:
+                return
+        except requests.ConnectionError:
+            pass
+
+        time.sleep(0.1)
+
+    raise TimeOutException()
+
+
+def test_expected_config(expectedconfig):
+    """
+    Configs as dumped by PostgREST should match an expected output.
+
+    Used to test default values, config aliases and environment variables. The
+    expected output for each file in 'configs', if available, is found in the
+    'configs/expected' directory.
+
+    """
+    expected = (basedir / "configs" / "expected" / expectedconfig).read_text()
+    assert dumpconfig(basedir / "configs" / expectedconfig) == expected
+
+
+def test_stable_config(configpath):
+    """
+    A dumped, re-read and re-dumped config should match the dumped config.
+
+    Note: only dump vs. re-dump must be equal, as the original config file might
+    be different because of default values, whitespace, and quoting.
+
+    """
+    env = {"ROLE_CLAIM_KEY": '."https://www.example.com/roles"[0].value'}
+    dumped = dumpconfig(configpath, moreenv=env)
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        tmpconfigpath = pathlib.Path(tmpdir, "config")
+        tmpconfigpath.write_text(dumped)
+        redumped = dumpconfig(tmpconfigpath, moreenv=env)
+
+    assert dumped == redumped
+
+
+def test_read_secret_from_file(secretpath):
+    if secretpath.suffix == ".b64":
+        configfile = basedir / "configs" / "base64-secret-from-file.config"
+    else:
+        configfile = basedir / "configs" / "secret-from-file.config"
+
+    secret = secretpath.read_bytes()
+
+    jwt = secretpath.with_suffix(".jwt").read_text()
+    headers = {"Authorization": f"Bearer {jwt}"}
+
+    with run(configfile, stdin=secret) as url:
+        response = requests.get(f"{url}/authors_only", headers=headers)
+        assert response.status_code == 200
+
+
+def test_read_dburi_from_file_without_eol():
+    with run(dburifromfileconfig, stdin=dburi) as url:
+        response = requests.get(f"{url}/")
+        assert response.status_code == 200
+
+
+def test_read_dburi_from_file_with_eol():
+    with run(dburifromfileconfig, stdin=dburi + b"\n") as url:
+        response = requests.get(f"{url}/")
+        assert response.status_code == 200
+
+
+def test_role_claim_key(roleclaimcase):
+    env = {"ROLE_CLAIM_KEY": roleclaimcase.key}
+    token = jwt.encode(roleclaimcase.data, secret).decode("utf-8")
+    headers = {"Authorization": f"Bearer {token}"}
+
+    with run(roleclaimkeyconfig, moreenv=env) as url:
+        response = requests.get(f"{url}/authors_only", headers=headers)
+        assert response.status_code == roleclaimcase.expected_status
+
+
+def test_invalid_role_claim_key(invalidroleclaimkey):
+    env = {"ROLE_CLAIM_KEY": invalidroleclaimkey}
+
+    with pytest.raises(subprocess.CalledProcessError):
+        dumpconfig(roleclaimkeyconfig, moreenv=env)
```

Comment:
is there any way we can show the dumped config here in case of a test failure?

i.e. when an incorrect role claim key does not throw an error and is parsed as correct - it would be helpful to see the dumped config, because this would show how the role claim key was parsed and dumped again.

Reply:
And maybe I can extend the `--dump-config` to do the following:

- `--dump-config` prints the whole file, just like it does now

- `--dump-config="jwt-role-claim-key"` prints just this one setting.

This would make the output in case of a failure here much more useful. It would also help with the boolean test-cases I have added on master. They dump the whole config file, but it's actually just about 3 values. They can easily be lost in the big picture.

WDYT?

Reply:
Fixed, this is what is printed when a role key is seen as valid:

```
========================================== FAILURES ===========================================
_____________________________ test_invalid_role_claim_key[.role] ______________________________

invalidroleclaimkey = '.role'

    @pytest.mark.parametrize("invalidroleclaimkey", FIXTURES["invalidroleclaimkeys"])
    def test_invalid_role_claim_key(invalidroleclaimkey):
        "Given an invalid role-claim-key, Postgrest should exit with a non-zero exit code."
        env = {"ROLE_CLAIM_KEY": invalidroleclaimkey}
                                               
        with pytest.raises(PostgrestError):
>           print(dumpconfig(CONFIGSDIR / "role-claim-key.config", moreenv=env))
E           Failed: DID NOT RAISE <class 'test_io.PostgrestError'>

test/io-tests/test_io.py:235: Failed
------------------------------------ Captured stdout call -------------------------------------
db-uri = "postgresql://postgres?host=/run/user/1001/tmp.iZohhdd0bV/socket&user=postgrest_test_authenticator"
db-schema = "test"                             
db-anon-role = "postgrest_test_anonymous"
db-pool = 1                                    
db-pool-timeout = 10                           
db-extra-search-path = "public"
db-channel = "pgrst"                           
db-channel-enabled = false
db-tx-end = "commit"                           
db-prepared-statements = true
server-host = "127.0.0.1"
server-port = 49421                            
server-unix-socket = ""
server-unix-socket-mode = "660"
openapi-server-proxy-uri = ""
jwt-secret = "reallyreallyreallyreallyverysafe" 
jwt-aud = ""                                   
secret-is-base64 = false
role-claim-key = ".\"role\""
max-rows = ""                                  
pre-request = ""                               
root-spec = ""                                 
raw-media-types = ""                           
log-level = "error"                            

=================================== short test summary info ===================================
FAILED test/io-tests/test_io.py::test_invalid_role_claim_key[.role] - Failed: DID NOT RAISE ...
================================ 1 failed, 43 passed in 12.66s ================================
```

Reply:
I think it's easier to add the filtering on the tests side - let me check

Reply:
I see what you mean. Have a look at those files, though:
- https://github.com/PostgREST/postgrest/blob/master/test/io-tests/configs/boolean-numeric.config
- https://github.com/PostgREST/postgrest/blob/master/test/io-tests/configs/expected/boolean-numeric.config

The only values I care about in the expected output are the lines that are given in the config. All the others are just default values and noise.

Maybe we can change the tests that test for "expected" config to make sure that all lines in the expectation file are there, but allow additional lines in the output?